File: ios/LegacyImport.swift
import Foundation

/// The pure half of the first-launch v9 import: which legacy rows to create.
///
/// Each v9 journal entry becomes a read-only settled row with key "legacy"
/// and id = the v9 upload id. Nothing is delivered: Diana reads the rows,
/// marks those transfers terminal, and cancels them. When the id also has a
/// v9 chunked manifest, the row reports its bytes, and the blob stays in the
/// directory until cancel(id).
///
/// A manifest with no journal entry makes no row. It stays dormant until a
/// same-id enqueue adopts it (a legacy row would be cancelled by Diana, and
/// the capture re-send needs the bytes).
enum LegacyImport {
  static let key = "legacy"
  static let fingerprint = "legacy"

  static func plan(events: [JournaledEventV9], manifests: [String: ChunkedManifestV9]) -> [QueueEntry] {
    // One row per id: the latest v9 outcome wins.
    var latest: [String: JournaledEventV9] = [:]
    for e in events where latest[e.id].map({ $0.timestamp <= e.timestamp }) ?? true {
      latest[e.id] = e
    }
    return latest.values.sorted { ($0.timestamp, $0.id) < ($1.timestamp, $1.id) }.compactMap { e in
      guard let state = state(e.type) else { return nil }
      let manifest = manifests[e.id]
      return QueueEntry(
        id: e.id, key: key, varsJSON: "null", descriptorJSON: "{}", url: nil, method: "POST",
        accept: [], retry: nil, bodyKind: .none,
        bodyPath: manifest == nil ? nil : ChunkedManifestV9.blobName,
        bodyContentType: nil, forceContentType: false, bodyFingerprint: fingerprint, parts: [],
        incarnation: manifest?.incarnation ?? UUID().uuidString, headers: [:], headerGeneration: 0,
        state: state, authParked: false, generation: 1, attempts: 0,
        bytesSent: manifest?.acceptedBytes ?? 0, totalBytes: manifest?.totalBytes ?? 0,
        expiresAt: manifest?.expiresAt ?? e.timestamp, nextAttemptAt: nil, settledEventId: nil,
        lastRequestId: nil, lastUrl: nil, lastPartIndex: e.partIndex, legacy: true,
        createdAt: e.timestamp, updatedAt: e.timestamp)
    }
  }

  static func state(_ v9Type: String) -> QueueEntry.State? {
    switch v9Type {
    case "completed": return .completed
    case "error": return .error
    case "cancelled": return .cancelled
    default: return nil
    }
  }
}

File: android/src/main/java/ai/openspace/backgroundupload/LegacyImport.kt
package ai.openspace.backgroundupload

import android.content.Context
import com.google.gson.Gson
import java.io.File

/**
 * First v10 launch. Each v9 journal entry becomes a read-only settled row
 * with key `legacy` and the v9 upload id. Nothing is delivered; the app
 * reads the rows with getRequests() and cancels them. v9 chunked manifests
 * are left in place: a same-id enqueue adopts them.
 *
 * The caller cancels the v9 WorkManager rows after this returns true. The
 * order is safe: under v10 code a v9 row exits at once (it has no entry
 * id), so nothing can write a v9 journal file after the import.
 *
 * Runs once, guarded by a marker file. A failed row save leaves the marker
 * unwritten, so the next launch tries again.
 */
object LegacyImport {
  const val MARKER = "v9-imported"
  const val KEY = "legacy"
  const val V9_JOURNAL_DIR = "rnbgupload-events"

  /** The v9 journal Entry, every field nullable: Gson reads whatever is there. */
  data class V9Entry(
    val eventId: String?,
    val uploadId: String?,
    val type: String?,
    val timestamp: Long?,
  )

  private val gson = Gson()

  /** Returns true when the import ran at this launch (no marker yet). */
  fun runOnce(context: Context, store: QueueStore): Boolean {
    val marker = File(QueueStore.rootDir(context), MARKER)
    if (marker.exists()) return false
    val complete = import(File(context.filesDir, V9_JOURNAL_DIR), store)
    if (complete) {
      runCatching { AtomicFiles.writeText(marker, "1") }
        .onFailure { Diag.error("could not write the v9 import marker", it) }
    }
    return true
  }

  /**
   * Imports every v9 journal entry in [v9Dir]. The newest entry per upload
   * id wins. Returns false when a row could not be saved (its file is kept).
   */
  internal fun import(v9Dir: File, store: QueueStore): Boolean {
    val files = v9Dir.listFiles { f -> f.extension == "json" } ?: return true
    val read = files.mapNotNull { f ->
      val entry = runCatching { gson.fromJson(f.readText(), V9Entry::class.java) }.getOrNull()
      if (entry == null) {
        Diag.warn("v9 journal file unreadable, skipped: ${f.name}")
        f.delete()
        null
      } else f to entry
    }
    var complete = true
    read.groupBy { it.second.uploadId }.forEach { (uploadId, group) ->
      val newest = group.maxByOrNull { it.second.timestamp ?: 0L }!!.second
      val row = if (uploadId == null) null else legacyRow(newest)
      val saved = when {
        row == null -> true
        store.load(row.id) != null -> true // a v10 entry already owns the id
        else -> runCatching { store.save(row) }.isSuccess
      }
      if (saved) group.forEach { it.first.delete() } else complete = false
    }
    return complete
  }

  internal fun legacyRow(entry: V9Entry): QueueEntry? {
    val id = entry.uploadId ?: return null
    val state = when (entry.type) {
      "completed" -> EntryState.COMPLETED
      "error" -> EntryState.ERROR
      "cancelled" -> EntryState.CANCELLED
      else -> return null
    }
    val at = entry.timestamp ?: 0L
    return QueueEntry(
      id = id,
      key = KEY,
      varsJson = "null",
      descriptor = null,
      body = null,
      state = state,
      attempts = 0,
      bytesSent = 0,
      totalBytes = 0,
      expiresAt = at,
      createdAt = at,
      updatedAt = at,
      generation = 1,
      legacy = true,
    )
  }
}
